tests/test_updater_delegation_graphs.py

```diff
@@ -251,16 +277,119 @@ def test_graph_traversal(self, test_data: DelegationsTestCase) -> None:
+        "unsigned delegated role": DelegationsTestCase(
+            delegations=[
+                TestDelegation("targets", "A"),
+                TestDelegation("targets", "B"),
+                TestDelegation("A", "C"),
+            ],
+            # The traversal stops after visiting an invalid role
+            visited_order=["A"],
+        )
+    }
+
+    @utils.run_sub_tests_with_dataset(invalid_metadata)
+    def test_invalid_metadata(self, test_data: DelegationsTestCase) -> None:
+        try:
+            self._init_repo(test_data)
+            # The invalid role is the last visited
+            invalid_role = test_data.visited_order[-1]
+            self.sim.signers[invalid_role].clear()
+
+            self.setup_subtest()
+            # The invalid role metadata must not be persisted
+            exp_files = [*TOP_LEVEL_ROLE_NAMES, *test_data.visited_order[:-1]]
+            exp_calls = [(role, 1) for role in test_data.visited_order]
+
+            updater = self._init_updater()
+            # Call explicitly refresh to simplify the expected_calls list
+            updater.refresh()
+            self.sim.fetch_tracker.metadata.clear()
+
+            with self.assertRaises(UnsignedMetadataError):
+                updater.get_targetinfo("missingpath")
+            # Check that there were no visited roles after the invalid one
+            # and only the valid metadata files were persisted
+            self.assertListEqual(self.sim.fetch_tracker.metadata, exp_calls)
+            self._assert_files_exist(exp_files)
+        finally:
+            self.teardown_subtest()
+
+
+class TestTargetFileSearch(TestDelegations):
+    r"""
+    Create a single repository with the following delegations:
+
+              targets
+         *.doc / \ r/*/*
+              A   B
+            r/x/* / \ r/y/*.zip
+                 C   D
+
+    Test that Updater successfully finds the target files metadata,
+    traversing the delegations as expected.
+    """
+
+    delegations_tree = DelegationsTestCase(
+        delegations=[
+            TestDelegation("targets", "A", paths=["*.doc", "*.md"]),
+            TestDelegation("targets", "B", paths=["releases/*/*"]),
+            TestDelegation("B", "C", paths=["releases/x/*"]),
+            TestDelegation("B", "D", paths=["releases/y/*.zip"]),
+        ],
+        target_files=[
+            TestTarget("targets", b"content", "targetfile"),
+            TestTarget("A", b"content", "README.md"),
+            TestTarget("C", b"content", "releases/x/x_v1"),
+            TestTarget("D", b"content", "releases/y/y_v1.zip"),
+        ],
+    )
+
+    def setUp(self) -> None:
+        super().setUp()
+        self._init_repo(self.delegations_tree)
+
+    targets: utils.DataSet = {
+        "targetfile": ("targetfile", []),
+        "README.md": ("README.md", ["A"]),
+        "releases/x/*": ("releases/x/x_v1", ["B", "C"]),
+        "releases/y/*.zip": ("releases/y/y_v1.zip", ["B", "D"]),
+    }
```

Comment:
Would it make sense to include boolean for whether the target is found or not? then you could include some negative tests here as well, using the same delegations_tree. Or do you want to keep them separate?

As an example 
* Add `TestTarget("C", b"x release in C", "releases/x/x_v2")`
* add test case `"target exists, not delegated": ("releases/x/x_v2", [ ... ], False)`

As another example
* Add `TestTarget("C", b"duplicate x release in C", "releases/x/x_v1")`
* existing test case "releases/x/*" now tests that it finds the correct target (and not the one added above). The case name could be improved to explain this

Reply:
I think extending it for negative cases makes sense but I didn't understand the second example. We add the target files in a dictionary in the simulator so the duplicate will replace the first target?

Reply:
oh right, if the target path is the same then the url is the same... you are correct this does not make sense.